### PR TITLE
feat: K3S Master API Whitelists

### PR DESCRIPTION
**Pull request by rayshaw001**
feat: K3S Master API Whitelists

### CODE REVIEW

**Comment by rayshaw001** [comment]
# Windows 下 WSL 2 主库main分支测试不通过
使用测试命令：`./test/integration_test.sh all`
测试结果如下：
```text
Job 'secretflow-psi' created successfully. You can use the following command to display job status:
  kubectl get kj
/home/kuscia
test_centralized_kuscia_api_http_available
ASSERT:KusciaApi healthZ http code expected:<200> but was:<000>
shunit2:ERROR test_centralized_kuscia_api_http_available() returned non-zero return code.
test_centralized_kuscia_api_grpc_available
Failed to dial target host "[172.18.0.2:8083](http://172.18.0.2:8083/)": context deadline exceeded
ASSERT:KusciaApi healthZ grpc status message expected:<success> but was:<>
shunit2:ERROR test_centralized_kuscia_api_grpc_available() returned non-zero return code.
kazi-kuscia-master
kazi-kuscia-lite-alice
kazi-kuscia-lite-bob
kazi-kuscia-master
kazi-kuscia-lite-alice
kazi-kuscia-lite-bob

Ran 3 tests.

FAILED (failures=2)
```

WSL 2 系统版本信息如下：
```bash
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.6 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.6 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

请教下可能是什么原因？

**Comment by rayshaw001** [comment]
# Ubuntu 22上 center.base测试通过
### 测试命令如下
```shell
make build
make image
export KUSCIA_IMAGE=docker.io/secretflow/kuscia:6ef6d50-20230826145247-6ef6d50
mkdir -p test
docker run --rm ${KUSCIA_IMAGE} cat /home/kuscia/scripts/test/integration_test.sh > ./test/integration_test.sh && chmod u+x ./test/integration_test.sh
./test/integration_test.sh center.base
```
image

**Comment by rayshaw001** [comment]
# p2p.base 测试失败，相关信息如下：
```bash
➜  kuscia git:(masterApiWhitelist) ✗ ./test/integration_test.sh p2p.base
➜  kuscia git:(masterApiWhitelist) ✗ docker exec -it kazi-kuscia-autonomy-alice bash
[root@kazi-kuscia-autonomy-alice kuscia]# scripts/user/create_example_job.sh PSI secretflow-psi-p2p-alice
/home/kuscia /home/kuscia
With JOB_EXAMPLE=PSI, job via APP_IMAGE=secretflow-image creating ...
kusciajob.kuscia.secretflow/secretflow-psi-p2p-alice created
Job 'secretflow-psi-p2p-alice' created successfully. You can use the following command to display job status:
  kubectl get kj
/home/kuscia
[root@kazi-kuscia-autonomy-alice kuscia]# kubectl get kj
NAME                       STARTTIME   COMPLETIONTIME   LASTRECONCILETIME   PHASE
secretflow-psi-p2p-alice   10s         10s              10s                 Failed
[root@kazi-kuscia-autonomy-alice kuscia]# 
[root@kazi-kuscia-autonomy-alice kuscia]# 
[root@kazi-kuscia-autonomy-alice kuscia]# kubectl describe kj secretflow-psi-p2p-alice
Name:         secretflow-psi-p2p-alice
Namespace:
Labels:       <none>
Annotations:  <none>
API Version:  kuscia.secretflow/v1alpha1
Kind:         KusciaJob
Metadata:
  Creation Timestamp:  2023-08-27T03:34:07Z
  Generation:          1
  Managed Fields:
    API Version:  kuscia.secretflow/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:initiator:
        f:maxParallelism:
        f:scheduleMode:
        f:stage:
        f:tasks:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2023-08-27T03:34:07Z
    API Version:  kuscia.secretflow/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:completionTime:
        f:lastReconcileTime:
        f:message:
        f:phase:
        f:reason:
        f:startTime:
    Manager:         kuscia
    Operation:       Update
    Subresource:     status
    Time:            2023-08-27T03:34:07Z
  Resource Version:  362
  UID:               d8fae63a-3caf-4dce-be65-b3b8b2967ef8
Spec:
  Initiator:        alice
  Max Parallelism:  2
  Schedule Mode:    BestEffort
  Stage:            Create
  Tasks:
    Alias:      single-psi
    App Image:  secretflow-image
    Parties:
      Domain ID:        alice
      Domain ID:        bob
    Priority:           100
    Task ID:            secretflow-psi-p2p-alice-single-psi
    Task Input Config:  {"sf_datasource_config":{"bob":{"id":"default-data-source"},"alice":{"id":"default-data-source"}},"sf_cluster_desc":{"parties":["alice","bob"],"devices":[{"name":"spu","type":"spu","parties":["alice","bob"],"config":"{\"runtime_config\":{\"protocol\":\"REF2K\",\"field\":\"FM64\"},\"link_desc\":{\"connect_retry_times\":60,\"connect_retry_interval_ms\":1000,\"brpc_channel_protocol\":\"http\",\"brpc_channel_connection_type\":\"pooled\",\"recv_timeout_ms\":1200000,\"http_timeout_ms\":1200000}}"},{"name":"heu","type":"heu","parties":["alice","bob"],"config":"{\"mode\": \"PHEU\", \"schema\": \"paillier\", \"key_size\": 2048}"}]},"sf_node_eval_param":{"domain":"preprocessing","name":"psi","version":"0.0.1","attr_paths":["input/receiver_input/key","input/sender_input/key","protocol","precheck_input","bucket_size","curve_type"],"attrs":[{"ss":["id1"]},{"ss":["id2"]},{"s":"ECDH_PSI_2PC"},{"b":true},{"i64":"1048576"},{"s":"CURVE_FOURQ"}],"inputs":[{"type":"sf.table.individual","meta":{"@type":"type.googleapis.com/secretflow.component.IndividualTable","schema":{"ids":["id1"],"features":["age","education","default","balance","housing","loan","day","duration","campaign","pdays","previous","job_blue-collar","job_entrepreneur","job_housemaid","job_management","job_retired","job_self-employed","job_services","job_student","job_technician","job_unemployed","marital_divorced","marital_married","marital_single"],"id_types":["str"],"feature_types":["float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float"]},"num_lines":"-1"},"data_refs":[{"uri":"alice.csv","party":"alice","format":"csv"}]},{"type":"sf.table.individual","meta":{"@type":"type.googleapis.com/secretflow.component.IndividualTable","schema":{"ids":["id2"],"features":["contact_cellular","contact_telephone","contact_unknown","month_apr","month_aug","month_dec","month_feb","month_jan","month_jul","month_jun","month_mar","month_may","month_nov","month_oct","month_sep","poutcome_failure","poutcome_other","poutcome_success","poutcome_unknown"],"labels":["y"],"id_types":["str"],"feature_types":["float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float"],"label_types":["int"]},"num_lines":"-1"},"data_refs":[{"uri":"bob.csv","party":"bob","format":"csv"}]}]},"sf_output_uris":["psi-output.csv"],"sf_output_ids":["psi-output"]}
    Tolerable:          false
Status:
  Completion Time:      2023-08-27T03:34:07Z
  Last Reconcile Time:  2023-08-27T03:34:07Z
  Message:              namespace "bob" not found
  Phase:                Failed
  Reason:               CheckJobTypeFailed
  Start Time:           2023-08-27T03:34:07Z
Events:
  Type     Reason           Age   From                       Message
  ----     ------           ----  ----                       -------
  Warning  KusciaJobFailed  24s   kuscia-controller-manager  KusciaJob failed to run
[root@kazi-kuscia-autonomy-alice kuscia]# 
[root@kazi-kuscia-autonomy-alice kuscia]# 
[root@kazi-kuscia-autonomy-alice kuscia]# 
[root@kazi-kuscia-autonomy-alice kuscia]# cat etc/kuscia.yaml
rootDir: /home/kuscia
domainID: alice
caKeyFile: etc/certs/ca.key
caFile: etc/certs/ca.crt
domainKeyFile: etc/certs/domain.key
master:
  endpoint:
  tls:
    certFile: /home/kuscia/etc/certs/domain.crt
    keyFile: /home/kuscia/etc/certs/domain.key
    caFile: /home/kuscia/etc/certs/master.ca.crt
externalTLS:
  certFile: /home/kuscia/etc/certs/external_tls.crt
  keyFile: /home/kuscia/etc/certs/external_tls.key
  caFile: /home/kuscia/etc/certs/ca.crt
agent:
  allowPrivileged: true
  plugins:
  - name: env-import
    config:
      usePodLabels: false
      envList:
      - envs:
        - name: system.transport
          value: transport.alice.svc
        - name: system.storage
          value: file:///home/kuscia/var/storage
        selectors:
        - key: maintainer
          value: secretflow-contact@service.alipay.com
  - name: config-render
```

**Comment by rayshaw001** [comment]
# 下面两个文件看起来需要一起修改下：
script/deploy/start_lite.sh
script/deploy/start_autonomy.sh

### script/deploy/start_autonomy.sh原文相关部分
```vim
 62 echo "rootDir: ${ROOT}
 63 domainID: ${NAMESPACE}
 64 caKeyFile: etc/certs/ca.key
 65 caFile: etc/certs/ca.crt
 66 domainKeyFile: etc/certs/domain.key
 67 master:
 68   endpoint: ${MASTER_ENDPOINT}
 69   tls:
 70     certFile: ${ROOT}/etc/certs/domain.crt
 71     keyFile: ${ROOT}/etc/certs/domain.key
 72     caFile: ${ROOT}/etc/certs/master.ca.crt
 73 externalTLS:
 74   certFile: ${EXTERNAL_TLS_CERT_FILE}
 75   keyFile: ${EXTERNAL_TLS_KEY_FILE}
 76   caFile: ${EXTERNAL_TLS_CA_FILE}
 77 agent:
 78   allowPrivileged: true
 79   plugins:
 80   - name: env-import
 81     config:
 82       usePodLabels: false
 83       envList:
 84       - envs:
 85         - name: system.transport
 86           value: transport.${NAMESPACE}.svc
 87         - name: system.storage
 88           value: file:///home/kuscia/var/storage
 89         selectors:
 90         - key: maintainer
 91           value: secretflow-contact@service.alipay.com
 92   - name: config-render
 93 " > etc/kuscia.yaml
```

### script/deploy/start_lite.sh原文相关部分
```vim
 47 echo "
 48 rootDir: ${ROOT}
 49 domainID: ${NAMESPACE}
 50 caKeyFile: etc/certs/ca.key
 51 caFile: etc/certs/ca.crt
 52 domainKeyFile: etc/certs/domain.key
 53 master:
 54   endpoint: ${MASTER_ENDPOINT}
 55   tls:
 56     certFile: ${ROOT}/etc/certs/domain.crt
 57     keyFile: ${ROOT}/etc/certs/domain.key
 58     caFile: ${ROOT}/etc/certs/master.ca.crt
 59 agent:
 60   allowPrivileged: ${ALLOW_PRIVILEGED}
 61   plugins:
 62   - name: config-render
 63   - name: env-import
 64     config:
 65       usePodLabels: false
 66       envList:
 67       - envs:
 68         - name: image
 69           value: secretflow-anolis8
 70         selectors:
 71         - key: maintainer
 72           value: secretflow-contact@service.alipay.com
 73 " > etc/kuscia.yaml
```

**Comment by rayshaw001** [comment]
> # 下面两个文件看起来需要一起修改下：
> script/deploy/start_lite.sh script/deploy/start_autonomy.sh
> 
> ### script/deploy/start_autonomy.sh原文相关部分
> ```viml
>  62 echo "rootDir: ${ROOT}
>  63 domainID: ${NAMESPACE}
>  64 caKeyFile: etc/certs/ca.key
>  65 caFile: etc/certs/ca.crt
>  66 domainKeyFile: etc/certs/domain.key
>  67 master:
>  68   endpoint: ${MASTER_ENDPOINT}
>  69   tls:
>  70     certFile: ${ROOT}/etc/certs/domain.crt
>  71     keyFile: ${ROOT}/etc/certs/domain.key
>  72     caFile: ${ROOT}/etc/certs/master.ca.crt
>  73 externalTLS:
>  74   certFile: ${EXTERNAL_TLS_CERT_FILE}
>  75   keyFile: ${EXTERNAL_TLS_KEY_FILE}
>  76   caFile: ${EXTERNAL_TLS_CA_FILE}
>  77 agent:
>  78   allowPrivileged: true
>  79   plugins:
>  80   - name: env-import
>  81     config:
>  82       usePodLabels: false
>  83       envList:
>  84       - envs:
>  85         - name: system.transport
>  86           value: transport.${NAMESPACE}.svc
>  87         - name: system.storage
>  88           value: file:///home/kuscia/var/storage
>  89         selectors:
>  90         - key: maintainer
>  91           value: secretflow-contact@service.alipay.com
>  92   - name: config-render
>  93 " > etc/kuscia.yaml
> ```
> 
> ### script/deploy/start_lite.sh原文相关部分
> ```viml
>  47 echo "
>  48 rootDir: ${ROOT}
>  49 domainID: ${NAMESPACE}
>  50 caKeyFile: etc/certs/ca.key
>  51 caFile: etc/certs/ca.crt
>  52 domainKeyFile: etc/certs/domain.key
>  53 master:
>  54   endpoint: ${MASTER_ENDPOINT}
>  55   tls:
>  56     certFile: ${ROOT}/etc/certs/domain.crt
>  57     keyFile: ${ROOT}/etc/certs/domain.key
>  58     caFile: ${ROOT}/etc/certs/master.ca.crt
>  59 agent:
>  60   allowPrivileged: ${ALLOW_PRIVILEGED}
>  61   plugins:
>  62   - name: config-render
>  63   - name: env-import
>  64     config:
>  65       usePodLabels: false
>  66       envList:
>  67       - envs:
>  68         - name: image
>  69           value: secretflow-anolis8
>  70         selectors:
>  71         - key: maintainer
>  72           value: secretflow-contact@service.alipay.com
>  73 " > etc/kuscia.yaml
> ```

# script/deploy/start_autonomy.sh 和 script/deploy/start_lite.sh加入如下内容后，测试全部通过！
```vim
 67 master:
 68   endpoint: ${MASTER_ENDPOINT}
 69   tls:
 70     certFile: ${ROOT}/etc/certs/domain.crt
 71     keyFile: ${ROOT}/etc/certs/domain.key
 72     caFile: ${ROOT}/etc/certs/master.ca.crt
 73   api-whitelist:
 74     - /(api(s)?(/[0-9A-Za-z_.-]+)?/v1(alpha1)?/namespaces/[0-9A-Za-z_.-]+/(pods|gateways|domainroutes|endpoints|services|events|c    onfigmaps|leases|taskresources|secrets|domaindatas|domaindatagrants|domaindatasources)(/[0-9A-Za-z_.-]+(/status$)?)?)
 75     - /api/v1/namespaces/[0-9A-Za-z_.-]+
 76     - /api/v1/nodes(/.*)?
```
image
image

**Comment by Eiji911** [comment]
start_autonomy.sh是需要加到脚本中去的，请加到脚本中去的。
start-lite不需要加的，但是代码需要修改。请见代码Review。

File: pkg/gateway/clusters/master.go [hunk 1]
```diff
@@ -182,10 +185,7 @@ func generateMasterInternalVirtualHost(cluster, service string, domains []string
 	if service == serviceAPIServer {
```
Comment (Eiji911) [inline pkg/gateway/clusters/master.go:185]: regex := getMasterApiWhitelistRegex(apiWhitelist)
if len(regex >0) {
virtualHost.Routes[0].Match.PathSpecifier = &route.RouteMatch_SafeRegex{
			SafeRegex: &matcherv3.RegexMatcher{
				Regex: "/(api(s)?(/[0-9A-Za-z_.-]+)?/v1(alpha1)?/namespaces/[0-9A-Za-z_.-]+" +
					"/(pods|gateways|domainroutes|endpoints|services|events|configmaps|leases|taskresources|secrets|domaindatas|domaindatagrants|domaindatasources)" +
					"(/[0-9A-Za-z_.-]+(/status$)?)?)|(/api/v1/namespaces/[0-9A-Za-z_.-]+)|(" +
					"/api/v1/nodes(/.*)?)",
				Regex: getMasterApiWhitelistRegex(apiWhitelist),
			},
		}
	}
}

**Comment by Eiji911** [comment]
Kuscia在部署脚本这块马上会有改动，并与当前PR会有冲突，要加快合入了。

File: pkg/gateway/clusters/master.go (hunk 1 above)
Comment (rayshaw001) [reply pkg/gateway/clusters/master.go:185]:
```go
if service == serviceAPIServer {
	regex := getMasterApiWhitelistRegex(apiWhitelist)
	if len(regex) > 0 {
		virtualHost.Routes[0].Match.PathSpecifier = &route.RouteMatch_SafeRegex{
			SafeRegex: &matcherv3.RegexMatcher{
				Regex: getMasterApiWhitelistRegex(apiWhitelist),
			},
		}
	}
}
```

Comment (rayshaw001) [reply pkg/gateway/clusters/master.go:185]: 已更新

**Comment by rayshaw001** [comment]
> start_autonomy.sh是需要加到脚本中去的，请加到脚本中去的。 start-lite不需要加的，但是代码需要修改。请见下面的Review。

已更新

**Comment by Eiji911** [comment]
> 
start_autonmy.sh要改的

**Comment by rayshaw001** [comment]
> 
改过了
image

**Review by Eiji911 — APPROVED** [review]
LGTM

**Review by longshan-ant — APPROVED** [review]
lgtm

**Pull request merged**